### PR TITLE
feat(Toast): Add Toast (extended from FloatingCard) GM-187

### PR DESCRIPTION
**Pull request by codecaaron**
### Overview

More pattern card floating goodness from https://github.com/Codecademy/client-modules/pull/1720

* Extends FloatingCard to add a structured Toast Component
* Adds Docs page for Toasts

Screen Shot 2021-06-09 at 8 40 30 AM

### PR Checklist

- [x] Related to designs: [Toast](https://www.figma.com/file/dT4E3SLymYjmhJFxbIOJz0/Toasts?node-id=0%3A1)
- [x] Related to JIRA ticket: GM-187
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

### CODE REVIEW

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.3.78-alpha.45bfa7.0<br/>@codecademy/gamut-labs@15.5.2-alpha.45bfa7.0<br/>@codecademy/gamut@32.11.2-alpha.45bfa7.0<br/>@codecademy/styleguide@36.13.4-alpha.45bfa7.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6092c9f003e7651dbc6398f5--gamut-preview.netlify.app>https://6092c9f003e7651dbc6398f5--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6092c9f003e7651dbc6398f5>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.13-alpha.d32118.0<br/>@codecademy/gamut-labs@19.0.6-alpha.d32118.0<br/>@codecademy/gamut@34.0.3-alpha.d32118.0<br/>@codecademy/styleguide@41.2.5-alpha.d32118.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c061bdc1e443b778f3ac57--gamut-preview.netlify.app>https://60c061bdc1e443b778f3ac57--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c061bdc1e443b778f3ac57>Deploy Logs</a></p>

**Review by dreamwasp — APPROVED** [review]
love it, new docs are  image

File: packages/gamut/src/Toast/Toast.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    'icon-message': {
+      gridTemplateAreas: `'icon message close'`,
+      minHeight: '104px',
+    },
+    'title-message': {
+      minHeight: '104px',
+      gridTemplateAreas: `'message message close'`,
+    },
+  },
+});
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/Toast/Toast.tsx:29]: umm, holy crap, this is beautiful??

File: packages/styleguide/stories/Molecules/Toast.stories.mdx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    children: 'I am a toast hear me roar!',
+  }}
+/>
+
+<br />
+
+### Use cases
+
+- Can be used when an immediate feedback is needed for learners upon completing a learning task. (E.g. unlocking an achievement during learning)
+- Should NOT repeat as a notification bell
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/stories/Molecules/Toast.stories.mdx:27]: Strong agree! ... but what does this mean, exactly? 😄

File: packages/styleguide/stories/Molecules/Toast.stories.mdx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    status: 'current',
+    figmaId: '',
+  }}
+  args={{
+    children: 'I am a toast hear me roar!',
+  }}
+/>
+
+<br />
+
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/stories/Molecules/Toast.stories.mdx:23]: Not a fan of different spacing across stories.. 

```suggestion
```

File: packages/styleguide/stories/Molecules/Toast.stories.mdx [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+### Use cases
+
+- Can be used when an immediate feedback is needed for learners upon completing a learning task. (E.g. unlocking an achievement during learning)
+- Should NOT repeat as a notification bell
+
+### Key Principles
+
+- Only one should appear at the screen at a time.
+- Toasts shouldn’t be distracting learning. Be cautious to use them in the Learning Environment, especially for messages that can be communicated outside of LE.
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/stories/Molecules/Toast.stories.mdx:32]: Nitty nits:

```suggestion
- Toasts shouldn’t distract from learning. Be hesitant to use them in the Learning Environment, especially for messages that can be communicated outside of the LE.
```

File: packages/styleguide/stories/Molecules/Toast.stories.mdx [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+### Style
+
+- Has a fixed width
+- Height changes depends on the body-text length.
+- Max. 3 rows of body-text is recommended.
+- Has variants, and colored different for light and dark theme.
+- Uses pattern at 80% opacity to give component depth.
+- Appears at the bottom right of the screen.
+- Slides in to the screen and when canceled, slides back and leaves the screen.
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/stories/Molecules/Toast.stories.mdx:42]: Are these two lines accurate? I would think them recommendations for usage, not core parts of the style?

File: packages/gamut/src/Toast/Toast.tsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    if (title) return 'title-message';
+    return 'message';
+  }, [title, icon]);
+
+  return (
+    <ToastContainer layout={layoutType} pattern="checkerDense">
+      {icon && (
+        <IconContainer alignSelf="center" backgroundImage={`url(${icon})`} />
+      )}
+      <Text gridArea="message" py={4}>
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/Toast/Toast.tsx:81]: Should this be a `Box`? It feels weird to have `Text` within `Text`.

Maybe we should add a custom lint rule...

**Review by JoshuaKGoldberg — APPROVED** [review]
A few questions / docs comments, but looks great overall!

File: packages/gamut/src/Toast/Toast.tsx (hunk 6 above)
Comment (codecaaron) [reply packages/gamut/src/Toast/Toast.tsx:81]: Good call out - Box definitely better here.

Text by default (for better or worse) is a span and isn't prescriptive about it styles until specified with something else. 
 This way we can switch colors - weight - font in side of other text elements semantically without having to match the styles or be super explicit.  `<Text as="h1">Some <Text textColor="yellow">Highlighted</Text></Text>`. 

Might not be the right approach for everything.
 
If we could check that nested `p` and `h1-6` tags aren't doing weird stuff I think that would be great.

File: packages/styleguide/stories/Molecules/Toast.stories.mdx (hunk 2 above)
Comment (codecaaron) [reply packages/styleguide/stories/Molecules/Toast.stories.mdx:27]: Not exactly sure what this meant but I updated it to something we can all agree on https://www.figma.com/file/dT4E3SLymYjmhJFxbIOJz0/Toasts?node-id=2%3A38

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.14-alpha.34d282.0<br/>@codecademy/gamut-labs@19.1.1-alpha.34d282.0<br/>@codecademy/gamut@34.0.3-alpha.34d282.0<br/>@codecademy/styleguide@41.3.1-alpha.34d282.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c14bc34c455432635c8261--gamut-preview.netlify.app>https://60c14bc34c455432635c8261--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c14bc34c455432635c8261>Deploy Logs</a></p>

**Pull request merged**